myia/compile/backends/python/python.py

```diff
@@ -0,0 +1,227 @@
+    # I guess this could change in the future ?
+    def make_arrow(self, user, graph):  # pragma: no cover
+        """Collect graph at module level."""
+        assert user is self, user
+        assert graph.parent is None
+        self.todo_graphs.append(graph)
+
+    def generate_directed_graphs(self, g):
+        """Collect and convert all graphs to directed graphs.
+
+        :param g: myia graph to compile.
+        :return: list of directed graphs, including given graph and all related module graphs.
+        """
+        self.todo_graphs.append(g)
+        seen_graphs = set()
+        directed_graphs = []
+        while self.todo_graphs:
+            graph = self.todo_graphs.pop()
+            if graph not in seen_graphs:
+                seen_graphs.add(graph)
+                directed_graphs.append(
+                    GraphToDirected(graph, self).generate_directed_graph()
+                )
+        return directed_graphs
+
+
+class PythonBackend:
+    """Python backend main class."""
+
+    def __init__(self, debug=False, pdb=False):
+        """Initialize.
+
+        :param debug: if False or None, do nothing.
+            If True, print generated code in stdout.
+            Otherwise, should be an output stream (e.g. stdout or a StringIO)
+            and generated code will be written into given stream.
+        :param pdb: if True, compiled function will be run in a pdb instance
+        """
+        if debug:
+            debug = sys.stdout if debug is True else debug
+            assert hasattr(debug, "write")
+
+        self.debug = debug
+        self.pdb = bool(pdb)
+
+    @classmethod
+    def nested_list_to_code_string(cls, structure, indentation=""):
+        """Convert a nested list of strings to a correctly indented Python code."""
+        code = ""
+        for index_entry, entry in enumerate(structure):
+            if not isinstance(entry, list):
+                code += f"{indentation}{entry}\n" if entry else "\n"
+            elif entry:
+                # Indent.
+                code += cls.nested_list_to_code_string(
+                    entry, indentation + "  "
+                )
+                # Add an extra-line if there is still code to write.
+                if (
+                    index_entry + 1 < len(structure)
+                    and structure[index_entry + 1]
+                ):
+                    code += "\n"
+        return code
+
+    def compile(self, graph):
+        """Compile given graph.
+
+        :param graph: myia graph to compile
+        :return: executable
+        """
+        code_generator = CodeGenerator()
+        code = []
+        for directed in GraphToModule().generate_directed_graphs(graph):
+            code.extend(code_generator.directed_graph_to_code(directed))
+
+        module = code_generator.globals
+        dynamic_imports = [
+            f"# Dynamic external import: {name}" for name in module
+        ]
```

Comment:
I think it would be possible to generate static imports as well, something like

```python
[f"from {value.__module__} import {value.__name__} as {name}" for name, value in module.items()]
```

That will only work if `value` has `__module__` and `__name__` and `value.__qualname__` does not contain dots, but I think that should be most of the time. The advantage is that this would allow us to write the code to a file to create a real module. In this case I think I would have methods `generate_code`, `write` and `compile` instead of just `compile`, so that the user can choose what they want to do.